validphys2/src/validphys/commondataparser.py

```diff
@@ -1,18 +1,310 @@
+        - shifts:
+            mapping with the single fktables and their respective shifts
+            necessary to create "gaps" that some cfactors or datasets might expect
+    """
+
+    repetition_flag: Optional[list[str]] = None
+    normalization: Optional[dict] = None
+    shifts: Optional[dict] = None
+
+
+@dataclasses.dataclass
+class TheoryMeta:
+    """Contains the necessary information to load the associated fktables"""
+
+    FK_tables: list[list]
+    operation: ValidOperation
+    conversion_factor: float = 1.0
+    comment: Optional[str] = None
+    apfelcomb: Optional[ValidApfelComb] = None
+    # The following options are transitional so that the yamldb can be used from the theory
+    appl: Optional[bool] = False
+    target_dataset: Optional[str] = None
+
+    def fktables_to_paths(self, grids_folder):
+        """Given a source for pineappl grids, constructs the lists of fktables
+        to be loaded"""
+        ret = []
+        for operand in self.FK_tables:
+            ret.append([grids_folder / f"{m}.{EXT}" for m in operand])
+        return ret
+
+    @classmethod
+    def parser(cls, meta: dict):
+        """The yaml databases in the server use "operands" instead of "FK_tables" """
+        if "operands" in meta:
+            meta["FK_tables"] = meta.pop("operands")
+        return parse_input(meta, cls)
+
+
+ValidTheory = Parser(TheoryMeta.parser)
+
+
+@dataclasses.dataclass
+class ValidKinematics:
+    """Contains all metadata for the kinematics of the dataset"""
+
+    file: ValidPath
+    variables: dict
+
+
+@dataclasses.dataclass
+class ValidReference:
+    """Holds literature information for the dataset"""
+
+    url: str
+    version: int = 0
+    tables: list[int] = dataclasses.field(default_factory=list)
+
+
+@dataclasses.dataclass
+class Variant:
+    """Defines the keys of the CommonMetaData that can be overwritten"""
+
+    data_uncertainties: list[ValidPath]
+
+
+@Parser
+def ValidVariants(variant_dict: dict) -> Dict[str, Variant]:
+    """Variants of a dataset are allowed to overwrite a subset of the keys of a dataset
+    (those defined in the Variant dataclass).
+    This wrapper class runs over the dictionary of variant and parses them into valid Variants
+    """
+    return {k: parse_input(v, Variant) for k, v in variant_dict.items()}
+
+
+@dataclasses.dataclass
+class CommonMetaData:
+    setname: str
+    ndata: int
+    observable: dict
+    kinematics: ValidKinematics
+    kinematic_coverage: dict
+    data_central: ValidPath
+    data_uncertainties: list[ValidPath]
+    dataset_label: str
+    plot_x: str
+    figure_by: list[str]
+    theory: ValidTheory
+    nnpdf_metadata: dict
+    version: int
+    version_comment: str = ""
+    arXiv: Optional[ValidReference] = None
+    iNSPIRE: Optional[ValidReference] = None
+    hepdata: Optional[ValidReference] = None
+    variants: Optional[ValidVariants] = dataclasses.field(default_factory=dict)
```

Comment:
This should not be optional if the default is not going to be None

Reply:
I can put the default to None of course. It is just a question of adding a `If variants is None` in a few places.

Reply:
Can as well make it a dict. But at the moment if you do set it to none it is broken in several places.